### ISSUE TITLE
Job updates for dockerized Cronicle 

### ISSUE THREAD

**Issue by wackazong** [jhuckaby/Cronicle#287]
## Summary

I have a dockerized Cronicle server that is running jobs on my containers with the shell plugin and ssh commands. The hostname of the cronicle container is set in my docker-compose to the hostname valid on the Docker network ("cronicle"). Problem is that I do not get any live job updates because the server retrieves its own hostname as "cronicle" and tries to connect from the browser to "cronicle" to get the updates. It should connect to cronicle.mydomain.com however.

Is there a way to get this connection working? Wouldn't it be better to use base_app_url as the url when connecting to the job progress websocket?

## Steps to reproduce the problem

Use a dockerized cronicle installation like https://hub.docker.com/r/intelliops/cronicle. Set the hostname in the docker-compose.yml file. Run a task. Check for live updates.

## Your Setup

Cronicle in a Docker container installed using https://hub.docker.com/r/intelliops/cronicle

### Operating system and version?

Windows 10 latest, Docker for Windows latest, latest Cronicle docker image

### Node.js version?

14.2

### Cronicle software version?

Version 0.8.46

### Are you using a multi-server setup, or just a single server?

Single

### Are you using the filesystem as back-end storage, or S3/Couchbase?

File system 

### Can you reproduce the crash consistently?

Yes

## Log Excerpts

Browser log:
https://cronicle:3013/socket.io/?EIO=3&transport=polling&t=N9FxvYx net::ERR_NAME_NOT_RESOLVED

Log in Cronicle Job view
Log Watcher: Server Connect Error: Error: xhr poll error (https://cronicle:3013)

**Comment by jhuckaby** [comment]
You'll have to forgive me, I've never used Docker, and I don't understand its unique requirements.  But it sounds like you need to have Cronicle connect via IP addresses instead of hostnames.  See the docs here:

### server_comm_use_hostnames

Setting this parameter to `true` will force the Cronicle servers to connect to each other using hostnames rather than LAN IP addresses.  This is mainly for special situations where your local server IP addresses may change, and you would prefer to rely on DNS instead.  The default is `false` (disabled), meaning connect using IP addresses.

### web_direct_connect

When this property is set to `false` (which is the default), the Cronicle Web UI will connect to whatever hostname/port is on the URL.  It is expected that this hostname/port will always resolve to your master server.  This is useful for single server setups, situations when your users do not have direct access to your Cronicle servers via their IPs or hostnames, or if you are running behind some kind of reverse proxy.

If you set this parameter to `true`, then the Cronicle web application will connect *directly* to your individual Cronicle servers.  This is more for multi-server configurations, especially when running behind a [load balancer](#load-balancers) with multiple backup servers.  The Web UI must always connect to the master server, so if you have multiple backup servers, it needs a direct connection.

Note that the ability to watch live logs for active jobs requires a direct web socket connection to the server running the job.  For that feature, this setting has no effect (it always attempts to connect directly).

### web_socket_use_hostnames

Setting this parameter to `true` will force Cronicle's Web UI to connect to the back-end servers using their hostnames rather than IP addresses.  This includes both AJAX API calls and Websocket streams.  You should only need to enable this in special situations where your users cannot access your servers via their LAN IPs, and you need to proxy them through a hostname (DNS) instead.  The default is `false` (disabled), meaning connect using IP addresses.

This property only takes effect if [web_direct_connect](#web_direct_connect) is also set to `true`.

**Comment by wackazong** [comment]
I found it. You have to set the hostname in the docker compose yaml file to the full domain name of the server. Then it works. Also you need to set the https port to 443 even if you are behind a reverse proxy because that is the port that the cronicle browser client will use to create a https connection to the server. I use hostnames and not IPs wherever possible and now it is working without a hitch.

**Comment by mikeTWC1984** [comment]
Yeah, live logs will not work with docker by default, because container will run in some virtual network and cronicle will use that internal ip by default, while user is hitting actual host name or localhost. If running docker on linux you can use host network for container (with --net=host). It could be more tricky if using reverse proxy for https. I think cronicle removes domain names from host name (although assuming above comment not sure about that now), but certificates are based of full name. I used custom_live_log_socket_url to overcome this but it requires some extra tweaks (https://github.com/jhuckaby/Cronicle/issues/313)

**Comment by LoSz** [comment]
> I found it. You have to set the hostname in the docker compose yaml file to the full domain name of the server. Then it works. Also you need to set the https port to 443 even if you are behind a reverse proxy because that is the port that the cronicle browser client will use to create a https connection to the server. I use hostnames and not IPs wherever possible and now it is working without a hitch.

@wackazong can you please share your docker-compose file. i cant figure out how to get it working
